clustering/clustering.py

```diff
@@ -0,0 +1,446 @@
+import os, sys
+from time import time
+import logging
+from django.conf import settings
+
+import clustering_settings as clust_settings
+
+# The following packages are only needed if the running process is configured to be a Celery worker. 
+# We avoid importing them in appservers to avoid having to install unneeded dependencies.
+if settings.IS_CELERY_WORKER:
+    from gaia_wrapper import GaiaWrapperClustering
+    import numpy as np
+    from sklearn import metrics
+    from sklearn.feature_selection import mutual_info_classif
+    import json
+    import networkx as nx
+    from networkx.readwrite import json_graph
+    import community as com
+    from networkx.algorithms.community import k_clique_communities, greedy_modularity_communities
+
+logger = logging.getLogger('clustering')
+
+
+class ClusteringEngine():
+    """Clustering Engine class.
+
+    This class regroups various methods for performing clustering on a set of sounds.
+
+    Instead of directly using the audio feature space for performing the clustering, it creates a 
+    K-Nearest Neighbors Graph that is then partitioned for obtaining the clusters.
+    It relies on Gaia for performing nearest neighbor searches on a multi-dimentional feature space
+    (in gaia_wrapper.py file). The available features used for clustering are listed in the 
+    clustering_settings.py file. It requires building the Gaia dataset index file in advanced.
+
+    It also includes some methods that enable to automaticaly estimate the performance of the clustering
+    method. Moreover, a few unsued alternative methods for performing some intermediate steps are left 
+    here for developement and research purpose.
+    """
+    def __init__(self):
+        self.gaia = GaiaWrapperClustering()
+
+    def _prepare_clustering_result_and_reference_features_for_evaluation(self, classes):
+        """Formats the clustering classes and some reference features in order to then estimate how good is the 
+        clustering performance. Tipically the reference features can be tag-derived features that reflect semantic 
+        characteristics of the content. The reference features are defined in the clustering settings file.
+        
+        Extracts reference features for the sounds given as keys of the classes argument.
+        Prepares classes and extracted features in lists in order to compare them.
+        Removes sounds with missing features.
+
+        Args:
+            classes (Dict{Int: Int}): clustering classes for each sound {<sound_id>: <class_idx>}.
+
+        Returns:
+            Tuple(List[List[Float]], List[Int]): 2-element tuple containing a list of evaluation features 
+                and list of classes (clusters) idx.
+        """
+        sound_ids_list, clusters = zip(*classes.items())
```

Comment:
could you just use classes.keys() and classes.values() here? I _think_ the ordering will be the same